.kokoro/trampoline_v2.sh

```diff
@@ -0,0 +1,352 @@
+    # bootstrapping secret.
+    SERVICE_ACCOUNT_KEY_FILE="${KOKORO_GFILE_DIR}/kokoro-trampoline.service-account.json"
+
+    mkdir -p "${tmpdir}/gcloud"
+    gcloud_config_dir="${tmpdir}/gcloud"
+
+    log_yellow "Using isolated gcloud config: ${gcloud_config_dir}."
+    export CLOUDSDK_CONFIG="${gcloud_config_dir}"
+
+    log_yellow "Using ${SERVICE_ACCOUNT_KEY_FILE} for authentication."
+    gcloud auth activate-service-account \
+	   --key-file "${SERVICE_ACCOUNT_KEY_FILE}"
+    gcloud auth configure-docker --quiet
+fi
+
+log_yellow "Changing to the project root: ${PROJECT_ROOT}."
+cd "${PROJECT_ROOT}"
+
+required_envvars=(
+    # The basic trampoline configurations.
+    "TRAMPOLINE_IMAGE"
+    "TRAMPOLINE_BUILD_FILE"
+)
+
+pass_down_envvars=(
+    # TRAMPOLINE_V2 variables.
+    # Tells scripts whether they are running as part of CI or not.
+    "RUNNING_IN_CI"
+    # Indicates which CI system we're in.
+    "TRAMPOLINE_CI"
+    # Indicates we're running trampoline_v2.
+    "TRAMPOLINE_V2"
+    # KOKORO dynamic variables.
+    "KOKORO_BUILD_NUMBER"
+    "KOKORO_BUILD_ID"
+    "KOKORO_JOB_NAME"
+    "KOKORO_GIT_COMMIT"
+    "KOKORO_GITHUB_COMMIT"
+    "KOKORO_GITHUB_PULL_REQUEST_NUMBER"
+    "KOKORO_GITHUB_PULL_REQUEST_COMMIT"
+    # For Build Cop Bot
+    "KOKORO_GITHUB_COMMIT_URL"
+    "KOKORO_GITHUB_PULL_REQUEST_URL"
+)
+
+if [[ -f "${PROJECT_ROOT}/.trampolinerc" ]]; then
+    source "${PROJECT_ROOT}/.trampolinerc"
+fi
+
+log_yellow "Checking environment variables."
+for e in "${required_envvars[@]}"
+do
+    if [[ -z "${!e:-}" ]]; then
+	log "Missing ${e} env var. Aborting."
+	exit 1
+    fi
+done
+
+log_yellow "Preparing Docker image."
+# Download the docker image specified by `TRAMPOLINE_IMAGE`
+
+set +e  # ignore error on docker operations
+# We may want to add --max-concurrent-downloads flag.
+
+log_yellow "Start pulling the Docker image: ${TRAMPOLINE_IMAGE}."
+if docker pull "${TRAMPOLINE_IMAGE}"; then
+    log_green "Finished pulling the Docker image: ${TRAMPOLINE_IMAGE}."
+    has_cache="true"
+else
+    log_red "Failed pulling the Docker image: ${TRAMPOLINE_IMAGE}."
+    has_cache="false"
+fi
+
+
+# The default user for a Docker container has uid 0 (root). To avoid
+# creating root-owned files in the build directory we tell docker to
+# use the current user ID.
+user_uid="$(id -u)"
+user_gid="$(id -g)"
+user_name="$(id -un)"
+
+# To allow docker in docker, we add the user to the docker group in
+# the host os.
+docker_gid=$(cut -d: -f3 < <(getent group docker))
+
+update_cache="false"
+if [[ "${TRAMPOLINE_DOCKERFILE:-none}" != "none" ]]; then
+    # Build the Docker image from the source.
+    context_dir=$(dirname "${TRAMPOLINE_DOCKERFILE}")
+    docker_build_flags=(
+	"-f" "${TRAMPOLINE_DOCKERFILE}"
+	"-t" "${TRAMPOLINE_IMAGE}"
+	"--build-arg" "UID=${user_uid}"
+	"--build-arg" "GID=${user_gid}"
+	"--build-arg" "USERNAME=${user_name}"
+	"--build-arg" "DOCKER_GID=${docker_gid}"
```

Comment:
Why set during build-time? Why can't these just be set at execution time?

Reply:
Sorry for a long explanation.

1. To allow sudo we need a real username
2. One of the tests (compute/oslogin iirc) needs a real username

So we need the real username for those reasons. So anyways we need to create a user and allow sudo at build time. Well, if we create a user, I can also add the user to the docker group.

Then we can pass `--user={$user_uid}:${user_gid}` which is better than `--user={$user_uid}:${docker_gid}`.

The files created by the former will be the same uid:gid as you, but the latter will create a file with different gid.

Usecase: we can use the script to generate the README:

```
scripts/run_tests_local.sh cdn readmegen
```

This will generate the README.rst in cdn directory with the same uid and gid as you.

Reply:
So basically, the direct answer to your question is:

Yes, we can pass `--user={$user_uid}:${docker_gid}` and docker works, but some other tests will fail (compute/oslogin, and potential tests which needs sudo access).

Reply:
Docker already has functionality for setting a [default user/group as part of a Dockerfile](https://docs.docker.com/engine/reference/builder/#user). What do we gain by having trampoline manage this instead of letting the container do it itself? 

I think my concern is that the container is supposed to be the encapsulation of the entire environment, including users/groups. I don't think it makes sense to have trampoline play a part in managing that. If folks want to have a different user/group by default, they should set that in their Dockerfile.

Reply:
The current code makes sure that, docker will run with the same uid:gid as the caller of the trampoline_v2.sh script. This is dynamic. When you run locally, it's going to be your usual uid:gid, when it's run in Kokoro, it's going to be uid:gid of kbuilder user, etc etc.

Reply:
So, to summarize:

1. I want to use `--user "${user_uid}:${user_gid}"` for making sure the container will create files with the same uid:gid as the caller.
2. It works for most of our tests, but some tests are failing.
3. compute/oslogin: We need a real user name for this uid -> we need to create the user in the container OS. This is why we passing the build arg.
4. docker in docker: We need to make sure the user in the container have a permission on /var/run/docker.sock. -> This is why we add our user to the group with the same id as the docker group in the host os.
5. sudo: Some tests potentially need root access, so we need to add sudoers entry on the fly. Actually our `run_tests.sh` needs it for storing cloudsql_proxy to /bin.

So there are reasons. If you come up with a solution with the passing tests both for `compute/oslogin` and `vision/automl`, I'm open to such a solution.

You can download this branch and try out the script with the following command:

```sh
$ scripts/decrypt-secrets.sh  # Make sure we have the secrets
$ scripts/run_tests_local.sh compute/oslogin py-3.7
$ scripts/run_tests_local.sh vision/automl py-3.7
# or
$ RUN_TESTS_DIRS=compute/oslogin:vision/automl \
  TRAMPOLINE_BUILD_FILE=.kokoro/tests/run_tests.sh \
  .kokoro/trampoline_v2.sh

```

These commands start from building the Docker image, so you can test end 2 end for your proposed solution.

Reply:
Doesn't this also mean that it rebuilds (and potentially pushes a new image) every time a new user runs it?

Reply:
Yes, but they are fast because `docker pull` can detect the local layers and Docker build is using `--cache-from`. These command never upload to anywhere by default. You have to provide two envvars to use that option like this:

```sh
TRAMPOLINE_IMAGE=gcr.io/tmatsuo-test/my-python-multi \
  TRAMPOLINE_IMAGE_UPLOAD=true
  scripts/run_tests_local.sh compute/oslogin py-3.7
```

Reply:
Having said that, the upload option is not intend to be used by humans.

My intention is to set that flag only in one of the periodic builds. When the tests are all green, it will push the built image. We won't need the Cloud Build hook any more (although it won't harm anything).

Reply:
I also hope that in the future, we can merge all the periodic builds into one. If we pass down the multiple nox sessions, it might be possible. The tests will take a very long time, so we may want to have a tool to run tests in parallel. I thought someone is developing such a tool.